### PR TITLE
Passkeys: Add support for "Log in with security keys"

### PR DESCRIPTION
**Pull request by momo-ozawa**
## Description
- Adds support for "log in with security keys" by removing `applinks:*.wordpress.com` from associated domains

## Notes
- I cherry-picked https://github.com/wordpress-mobile/WordPress-iOS/pull/21739/commits/e62fbeeb50a7b488976dd2400f59bbb4bb0dca7c (Thanks @Ecarrion) 
- Removing `applinks:*.wordpress.com` from associated domains will disable deep linking into the Reader for `*.wordpress.com/:post_year/:post_month/:post_day/:post_name` links (ref: https://github.com/wordpress-mobile/WordPress-iOS/pull/16319)
- We've decided that removing deep linking support for `*.wordpress.com/:post_year/:post_month/:post_day/:post_name` links is an acceptable trade-off to get passkeys working (Slack ref: p1699395576443799/1698884760.893009-slack-C0180B5PRJ4)

## How to test
- Log into an account that has security keys enabled using your username and password
- In the 2FA step, select "Use a security key"
- Verify you're able to log in

https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/aa9ede60-5e0a-4a27-a98d-b4755cfef57e

## Regression Notes
1. Potential unintended areas of impact
Log in, Deep linking into Reader for `*.wordpress.com` links

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually for log in, reached a decision that removing support for `*.wordpress.com` deep linking was an acceptable trade-off (see notes above)

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/WordPressDebug.entitlements [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		<string>Default</string>
+	</array>
+	<key>com.apple.developer.associated-domains</key>
+	<array>
+		<string>webcredentials:wordpress.com</string>
+		<string>applinks:wordpress.com</string>
+		<string>applinks:apps.wordpress.com</string>
+		<string>applinks:public-api.wordpress.com</string>
+		<string>applinks:links.wp.a8cmail.com</string>
+		<string>applinks:*.wordpress.com</string>
```
Comment (guarani) [inline WordPress/WordPressDebug.entitlements]: 👋 @momo-ozawa, is this intentional? We've removed it for the Jetpack target's entitlement files above to allow passkeys to work, so I'm not sure if this was added here on purpose or not.

File: WordPress/WordPressDebug.entitlements
Comment (guarani) [inline WordPress/WordPressDebug.entitlements:1]: It seems like this new entitlements file isn't used:
Screenshot 2023-11-08 at 10 07 15

Is this new file needed? I don't see a need for it but I could be missing something as I haven't tested this PR on WordPress yet, just commenting now to leave early feedback.

**Comment by kean** [comment]
I re-tested today using the build from this PR and my work profile. The observations are the same as before:

- I doesn't work for me and this time even restarting the iPhone didn't help
- The 2FA with an OTP fails with a "not a valid two-factor authentication code error"

File: WordPress/WordPressDebug.entitlements
Comment (momo-ozawa) [reply WordPress/WordPressDebug.entitlements:1]: Good call, removed! https://github.com/wordpress-mobile/WordPress-iOS/pull/22001/commits/78657f860101a168f689d027d8f84253ee1c81e2

**Comment by momo-ozawa** [comment]
@kean Thanks for reviewing!

> I doesn't work for me and this time even restarting the iPhone didn't help

Hmmm! If I log in via username / password, I can use my Passkey saved on my iPhone during the "Use a security key" step. I've tested on both JP and WP debug builds, and can log in via this method

> The 2FA with an OTP fails with a "not a valid two-factor authentication code error"

I believe this is expected behavior. Once security keys are set, I don't think you can complete 2FA via OTP. (Slack ref: p1699376470234739/1698884760.893009-slack-C0180B5PRJ4)

> (New) I managed to get an infinite spinner after failing the login, going to the root screen, and trying again. Update: I an no longer login and see an infinite/long spinner even after restating and reinstalling the app :(

Is this reproducible? I tried the following but couldn't reproduce
- Log in via username / password
- Use OTP for 2FA step
- Go back to root screen
- Log in via username / password
- User security keys for 2FA step, successfully logged in

**Comment by kean** [comment]
> I believe this is expected behavior. 

I just pointed it out because the error message does not indicate this. We are risking getting a lot of complains from A8C users.

> Is this reproducible? 

Not anymore as I deleted the security key.

**Comment by guarani** [comment]
> I believe this is expected behavior. Once security keys are set, I don't think you can complete 2FA via OTP.

If so, the app could avoid showing this screen altogether to avoid asking for an OTP code that it can't accept. In my testing of WP.com in the browser ([internal ref](p1699380078642219/1698884760.893009-slack-C0180B5PRJ4)), I can use either of them to log in when both are enabled on my WP.com account. This suggests that end users who enable passkeys should still be able to log in using their OTP. 

Do you think OTP should work even with passkeys enabled, or should we go with the updated error message you mention in p1699461145683579/1698884760.893009-slack-C0180B5PRJ4 @momo-ozawa?

**Comment by guarani** [comment]
> * _Update_: I an no longer login and see an infinite/long spinner even after restating and reinstalling the app :(

Did you try restarting the device and re-installing the app? Definitely odd that it gets stuck like this.

**Comment by guarani** [comment]
> * _Update_: I an no longer login and see an infinite/long spinner even after restating and reinstalling the app :(

The spinner was shown for 16 seconds on my device, that's how long it took to log in. How long did you wait? Logging into a work profile as you mention might take much longer than a regular test account with few sites.

**Review by guarani — APPROVED** [review]
Sharing my happy path testing here:

### WordPress app

- [x] Logged in with passkey on WordPress app
- [x] Verified Reader links still open the WordPress app to the Reader tab (e.g. `https://wordpress.com/read/*`)
- [x] Verified blog links no longer open the Reader tab (e.g. `*.wordpress.com/:post_year/:post_month/:post_day/:post_name`)


### Jetpack app

- [x] Logged in with passkey on Jetpack app
- [x] Verified Reader links still deep link to the post in the Reader (e.g. `https://wordpress.com/read/*`)
- [x] Verified blog links no longer open the Reader tab (e.g. `*.wordpress.com/:post_year/:post_month/:post_day/:post_name`)

I think there are a couple of things to clarify (not blockers since this PR is a clear improvement over what's already in the beta):
- How are we handling users who have passkey support but wish to log in via their OTP code? Do we support this, as is the case on WP.com, or will this be unsupported on the app? Some users may create a passkey in their browser and not sync it to their device, so maintaining OTP support for these users might be very valuable (if it's permitted for their account)
- If a user is required to use a passkey, should we hide the screen that asks for an OTP? This is how the flow would look: enter your email & password then the app opens the system bottom sheet with the "Sign in with your passkey?" prompt
- If we haven't already, I think additional testing based on p91TBi-aHu-p2 is needed

**Comment by kean** [comment]
> The spinner was shown for 16 seconds on my device, that's how long it took to log in. How long did you wait? Logging into a work profile as you mention might take much longer than a regular test account with few sites.

Normally, the login takes around a second or two for me. I didn't time it, but I waited for minutes.

**Comment by enejb** [comment]
I tested the following. 

In Jetpack. For a non automattician
[x] Latest 23.5.2 
If you have auth key enabled. Non sms. 
We still show the user with the SMS option. Which is a bit confusing since once they click on the send sms link they get a confirmation that send tells them that an sms was sent which is not the case at all :( 

I am able to login via the auth codes provided by the auth app. 

[x] This PR. 
If you have auth key enabled. Non sms. 
Same as the latest version. I think this should be addressed in future PRs. 

I am able to login via the auth codes provided by the auth app.  (this use doesn't have passkeys setup)

If you have passkey setup I am not able to login via the auth app part.

**Comment by guarani** [comment]
> We still show the user with the SMS option. Which is a bit confusing since once they click on the send sms link they get a confirmation that send tells them that an sms was sent which is not the case at all :(

Recapping what @enejb and I just discussed offline, this might be fixable once the changes in https://github.com/wordpress-mobile/WordPressKit-iOS/pull/646 land. We can follow up on it over in https://github.com/wordpress-mobile/WordPress-Android/issues/11746.

**Comment by guarani** [comment]
It looks a small change is needed here in the WPiOS codebase to get non-passkey accounts to log in due to it expecting "two_step_nonce_webauthn" in this scenario. I'll need to loop back to investigate.

**Comment by momo-ozawa** [comment]
@guarani Thanks for the WPKit PR! Confirming that I can now log in via authenticator OTP on my a8c account, and I see authenticator in the list of supported auth types.

```
        "two_step_supported_auth_types" =         (
            backup,
            authenticator,
            push,
            webauthn
        );
```

**Comment by guarani** [comment]
> Confirming that I can now log in via authenticator OTP on my a8c account, and I see authenticator in the list of supported auth types.

Thanks for checking! I just pushed a change to fix login for accounts without passkeys that have 2FA enabled.

**Comment by guarani** [comment]
Tested 9b0913ae0c40dcf5b43b345e59040f33ad07b21c with these types of login:

### WP.com login scenarios
- [x] Passkey and app-based 2FA, user chooses passkey
- [x] Passkey and app-based 2FA, user chooses app-based 2FA
- [x] App-based 2FA (no passkey configured)
- [x] SMS-based 2FA (no passkey configured) 
  - Here the app offers me a security key, offers to send SMS as an alternative (when that's the account's default) and prompts me for an app-based 2FA code which the account is not configured to have. This is an existing problem as @enejb points out in https://github.com/wordpress-mobile/WordPress-iOS/pull/22001#issuecomment-1802558422, but it's exacerbated by the addition of a new authentication method, passkeys. A good solution would be to only show 2FA options that the user has, so if they don't have a passkey, we don't show it to the user as an option. If they don't have SMS, we shouldn't show it, etc. 
- [x] No password set, magic link login  

### Social login scenarios
- [x] Google Social login 
- [x] Apple Social login
- [x] Google Social login with 2FA enabled
- [x] Apple Social login with 2FA enabled


### WP.com signup scenarios
- [x] New sign up

### Other
- [x] Account closure

**Comment by momo-ozawa** [comment]
Some flows not covered above that aren't working as expected that I'm looking into:
- When I exit out of the security keys prompt, I see an error message saying "Whoops, something went wrong". I'd expect not to see any error message
  - Update: I've pushed changes to https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/801 and am pointing to this branch in https://github.com/wordpress-mobile/WordPress-iOS/pull/22001/commits/e8851f307e65e2267819e0644f4dfbeefc418f4e
- ~When I try logging in an SMS only account, I get an error "The operation couldn't be completed"~

**Comment by guarani** [comment]
✅ Tested 92d73c3e15d0f439a1e01db40e575f82c11f4595 with these types of login

### WP.com login scenarios
- [x] Passkey and app-based 2FA, user chooses passkey
- [x] Passkey and app-based 2FA, user chooses app-based 2FA
- [x] App-based 2FA (no passkey configured)
- [x] No password set, magic link login  

### Social login scenarios
- [x] Google Social login 
- [x] Apple Social login
- [x] Google Social login with 2FA enabled
- [x] Apple Social login with 2FA enabled


### WP.com signup scenarios
- [x] New sign up

### Other
- [x] Account closure

**Comment by guarani** [comment]
👋 @mokagio, this is ready to be included in the beta fix 🙇

**Comment by mokagio** [comment]
@guarani noted. By the way, feel free to merge the PR next time 😄 Everyone should have the user rights to merge PRs that fulfill the branch protection rules in the release branch and our automation always fetches the latest changes before building a new binary.

**Comment by guarani** [comment]
> By the way, feel free to merge the PR next time

Gotcha! Noted for next time 👍  I wasn't sure if the release branch was a special case.

**Pull request merged**